### ISSUE TITLE
Getting message "Please authorize client c..."

### ISSUE THREAD

**Issue by 19Norseman** [simonrob/email-oauth2-proxy#70]
My thanks in advance to simonrob and other contributors. This is a nice piece of work,

I'm pretty sure this in an issue either in my emailproxy.config or in the way I registered my app with Azure.
**I don't think its an issue with emailproxy.py.**
I'm looking for some help with someone who has been through this.
If you have been through this could you provide a step by step for registering the app with Azure.
I'm pretty sure the issue is in the app permissions and/or accounts that can use app.
When registering I specified "Single tenant", ie. only accounts within this tenant.
Because I registered under cervocraig@gmail.com and I'm trying to connect under ccervo@alum.mit.edu I suspect this might be the issue, but don't know how to fix it.

1. I Running on Ubuntu 20.04
2. I've built a custom app that sends email via SMTP using basic authentication'
3.  I've downloaded email-oauth2-proxy-main.zip
4. I unpacked it and I edited the config file as described in README.md
5. I logged onto AZURE using a Google/gmail account I have.
6. Went through the process of registering an app "gettemp3"
7. I have the tenant id, app id and app secret and I entered them in the emailproxy.config.
8. I can success fully run emailproxy.py
9. For debugging purposes I am using telnet to connect to the proxy (pop) and I'm manually responding to via telnet.
10. I get the following log from the proxy:
2022-10-03 17:55:28: Initialising Email OAuth 2.0 Proxy from config file /home/ccervo/Desktop/OAuth2Proxy/email-oauth2-proxy-main/emailproxy.config
2022-10-03 17:55:28: Starting IMAP server at localhost:1993 (unsecured) proxying outlook.office365.com:993 (SSL/TLS)
2022-10-03 17:55:28: Starting POP server at localhost:1995 (unsecured) proxying outlook.office365.com:995 (SSL/TLS)
2022-10-03 17:55:28: Starting SMTP server at localhost:1587 (unsecured) proxying smtp.office365.com:587 (STARTTLS)
2022-10-03 17:55:28: Starting IMAP server at localhost:2993 (unsecured) proxying imap.gmail.com:993 (SSL/TLS)
2022-10-03 17:55:28: Starting POP server at localhost:2995 (unsecured) proxying pop.gmail.com:995 (SSL/TLS)
2022-10-03 17:55:28: Starting SMTP server at localhost:2465 (unsecured) proxying smtp.gmail.com:465 (SSL/TLS)
2022-10-03 17:55:28: Initialised Email OAuth 2.0 Proxy - listening for authentication requests
2022-10-03 17:55:35: New incoming connection to POP server at localhost:1995 (unsecured) proxying outlook.office365.com:995 (SSL/TLS)
2022-10-03 17:55:35: Accepting new connection to POP server at localhost:1995 (unsecured) proxying outlook.office365.com:995 (SSL/TLS) via ('127.0.0.1', 60610)
2022-10-03 17:55:35: POP (localhost:1995; 127.0.0.1:60610->outlook.office365.com:995) --> [ Client connected ]
2022-10-03 17:55:35: POP (localhost:1995; 127.0.0.1:60610->outlook.office365.com:995) <-> [ Starting TLS handshake ]
2022-10-03 17:55:35: POP (localhost:1995; 127.0.0.1:60610->outlook.office365.com:995) <-> [ TLS handshake complete ]
2022-10-03 17:55:35: POP (localhost:1995; 127.0.0.1:60610->outlook.office365.com:995)     <-- b'+OK The Microsoft Exchange POP3 service is ready. [TQBOADIAUABSADIAMABDAEEAMAAwADMANgAuAG4AYQBtAHAAcgBkADIAMAAuAHAAcgBvAGQALgBvAHUAdABsAG8AbwBrAC4AYwBvAG0A]\r\n'
2022-10-03 17:55:35: POP (localhost:1995; 127.0.0.1:60610->outlook.office365.com:995) <-- b'+OK The Microsoft Exchange POP3 service is ready. [TQBOADIAUABSADIAMABDAEEAMAAwADMANgAuAG4AYQBtAHAAcgBkADIAMAAuAHAAcgBvAGQALgBvAHUAdABsAG8AbwBrAC4AYwBvAG0A]\r\n'
2022-10-03 17:56:01: POP (localhost:1995; 127.0.0.1:60610->outlook.office365.com:995) --> b'User ccervo@alum.mit.edu\r\n'
2022-10-03 17:56:01: POP (localhost:1995; 127.0.0.1:60610->outlook.office365.com:995) <-- b'+OK\r\n'
2022-10-03 17:56:26: POP (localhost:1995; 127.0.0.1:60610->outlook.office365.com:995) --> b'PASS [[ Credentials removed from proxy log ]]\r\n'
2022-10-03 17:56:26: POP (localhost:1995; 127.0.0.1:60610->outlook.office365.com:995)     --> b'AUTH XOAUTH2\r\n'
2022-10-03 17:56:26: POP (localhost:1995; 127.0.0.1:60610->outlook.office365.com:995)     <-- b'+ \r\n'
2022-10-03 17:56:26: Authorisation request received for ccervo@alum.mit.edu (interactive mode)

This is the telnet log:
$ telnet localhost 1995
Trying 127.0.0.1...
Connected to localhost.
Escape character is '^]'.
+OK The Microsoft Exchange POP3 service is ready. [TQBOADIAUABSADIAMABDAEEAMAAwADMANgAuAG4AYQBtAHAAcgBkADIAMAAuAHAAcgBvAGQALgBvAHUAdABsAG8AbwBrAC4AYwBvAG0A]
User ccervo@alum.mit.edu
+OK
PASS ????????? (credentials hidden)
-ERR Authentication failed. Email OAuth 2.0 Proxy: Login failed - the authentication request expired or was cancelled for account ccervo@alum.mit.edu

When I enter "PASS ?????????"
I get a pop-up with emailproxy icon and pulldown and the message
"Please authorize client c...."
Before I do anything the pop up disappears.
The proxy is hung I have to control C out of it and then I get the following in the telnet window.
"-ERR Authentication failed. Email OAuth 2.0 Proxy: Login failed - the authentication request expired or was cancelled for account ccervo@alum.mit.edu"

**Comment by simonrob** [comment]
The proxy pauses the login to allow you to continue the OAuth process in a web browser. Once you see the notification, click on the proxy's icon and select your account in the authorisation submenu. You will then get a prompt to log in and authorise the connection (which is essentially a one-off action).

**Comment by 19Norseman** [comment]
The message and icon disappear before I can click it. Never see a browser.

Craig   Cervo
19 Norseman Ave
Glouster, MA 01939
978-394-5280
***@***.***
Phone
________________________________

**Comment by D4rki91** [comment]
Hi,

Im not sure if this is helping you.
But I had also the problem, that my Proxy was crashing during the external-auth process. (Windows Server 2019)
I had to install `"pip install --pre pythonnet" and"pip install pythonnet==3.0.0a2"` to make it work properly.

**Comment by 19Norseman** [comment]
Thanks, I'll try that. I'd try to debug this, but no Python experience.

Craig   Cervo
19 Norseman Ave
Glouster, MA 01939
978-394-5280
***@***.***
Phone
________________________________

**Comment by 19Norseman** [comment]
I found my problem. It was **operator error**.
While following the instructions in README.md I was confused by
  _"The first time your email client makes a request you should see a notification from the proxy about authorising your account.
 (Note that the notification is not itself clickable, but pull requests to improve this are very welcome)."_
and 
_"Click the proxy's menu bar icon."_

**I was clicking on the icon on the notification, not the icon on the menu bar.**
The icon the menu bar is so small I hadn't noticed it.
Once I realized the mistake and clicked on the menu bar icon I made progress.
I got the pop-up window asking me to authorize my client.
I filled in my user name and password and clicked login.

**New problem:**
 I get a pop-up window from Azure saying 
_"This app may be risky. If you trust this app please ask your admin to grant access."_
It also says
_"Have an admin account? Sign in with that account"_

I'm not sure what to do next.
Anyone familiar with registering apps on Azure that can help me out.

My current status with Azure.
I created an Azure account using my cervocraig@gmail account.
I enrolled in a "free trial subscription".
I registered my app "gettemp3" and filled in the information using my best judgement.

emailproxy.py seems to be working as described, but I'm blocked "authorizing my account" by this admin thing.

**Comment by D4rki91** [comment]
Hey,

sure I can help you with that. I will provide you the screenshots of the "manual" I've written for my company.

First you have to navigate to Azure AD: https://aad.portal.azure.com

![image](https://user-images.githubusercontent.com/107997738/193909184-9b510ab7-92fe-4502-b5af-e5df68f09df2.png)

Click on App registrations.

Now register a New Application as following.

![image](https://user-images.githubusercontent.com/107997738/193909790-222ba6ce-b3a5-41bf-ae34-8456318559b4.png)

Now go to “Certificates & secrets” and add a new secret.

![image](https://user-images.githubusercontent.com/107997738/193909971-80581358-546d-428d-bf15-b79dd33fbf87.png)

The secret can be valid for a maximum of 24 Months before you must renew it.

![image](https://user-images.githubusercontent.com/107997738/193910026-d5efeec9-af90-467c-8564-185a355414eb.png)

Now we have the Client Secret copy and save the “Value”. The Value is the Client Key the Proxy Application will need to authenticate the user mailboxes afterwards.

The next step is to configure the API permissions.

![image](https://user-images.githubusercontent.com/107997738/193910104-41a74c8f-40fc-4d52-9590-63131eb4142d.png)

Click on Add permission.
Select the Microsoft Graph

![image](https://user-images.githubusercontent.com/107997738/193910137-b1fab668-d524-425b-9fc2-e3291c335186.png)

Now add the following “delegated” permissions, you can use the search bar and checkmark multiple entries.

![image](https://user-images.githubusercontent.com/107997738/193910207-b8354d94-e67b-4a47-8698-dc260418908b.png)
![image](https://user-images.githubusercontent.com/107997738/193910226-44c4f839-44b7-49c1-9654-5adff5c1ae50.png)

You should grant admin consent to all the permissions.

![image](https://user-images.githubusercontent.com/107997738/193910315-dc54d303-7452-47c1-94da-dd50aff66a7e.png)

We also need the Application ID which can be found here:

![image](https://user-images.githubusercontent.com/107997738/193910390-6b545032-3a61-4683-a78d-d75b821fb28c.png)

**Comment by D4rki91** [comment]
I think you were missing the step where you have to "Grant admin constent for your tennant".

**Comment by 19Norseman** [comment]
Thanks. Much appreciated.

Craig   Cervo
19 Norseman Ave
Glouster, MA 01939
978-394-5280
***@***.***
Phone
________________________________

**Comment by simonrob** [comment]
Thanks @D4rki91 for providing this useful guide to AAD client setup.

@19Norseman if you are not able to get your client past admin consent, it may be worth trying an existing client ID and secret from one of the various open source email clients that are often pre-authorised on many email servers – see the proxy's [readme](https://github.com/simonrob/email-oauth2-proxy/blob/main/README.md) for details.

**Comment by 19Norseman** [comment]
Very good suggestion. I'll try it. I'm pretty sure all is well with the proxy.

 Question, because I'm trying to connect to an M.I.T. email account, I assume I need to get the M.I.T. admin to give admin consent. Is that right? I'm confused over who's admjn consent I need.

Craig   Cervo
19 Norseman Ave
Glouster, MA 01939
978-394-5280
***@***.***
Phone
________________________________

**Comment by simonrob** [comment]
Yes, that's right. From experience, though, the prompt at authorisation time to notify the administrator about a consent request rarely works, so you may need to request this manually.

**Comment by 19Norseman** [comment]
Thanks.

Craig   Cervo
19 Norseman Ave
Glouster, MA 01939
978-394-5280
***@***.***
Phone
________________________________

**Comment by 19Norseman** [comment]
Once again many thanks for your patience and help.
So SimonRob, I followed your advice.
I used the client ID and client secret from https://github.com/M66B/FairEmail/blob/master/app/src/main/res/xml/providers.xml
First attempt failed because **I didn't reset redirect_uri** in config to redirectUri="https://oauth.faircode.eu/"
Second attempt failed. **I tried to use SMTP and got error** "Client not authenticated to send mail."
Third attempt succeeded. **I used ESMTP format instead of SMTP.**
So I have verified that my download of emailproxy.py was successful and emailproxy.py is working as designed.

I was motivated to go down this path because I have a custom built service running on an Ubuntu server.
The service was using my account @alum.mit.edu to send alarm messages to me for various error conditions.
The service uses SMTP and basic authentication.
**On 10/1/2022 MS and MIT disabled basic authentication for SMTP.**
This caused my service to fail.
Note MS claims raison d'être for doing this was "improved security".
**That's garbage!** It was clearly to improve their monopoly.
Your proxy and the existence of previously authorized credentials free available on the web clearly demonstrate there is no improvement in security.

I'm going to pursue completing the registration of my app on Azure.
I'm not inclined to finance MS by paying for a subscription on Azure, so when my "free trial" expires I'll scrum the web for free credentials.

Once again thanks for your patience and help.
I'll close this issue.
